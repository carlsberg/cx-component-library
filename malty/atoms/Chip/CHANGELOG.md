# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- **button, chip:** Remove haveStyle checks ([#708](https://github.com/CarlsbergGBS/cx-component-library/issues/708)) ([5c9762d](https://github.com/CarlsbergGBS/cx-component-library/commit/5c9762d63041ad967e7b8390d6fa32c59671a64c))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **chip:** [DSM-700] Add tests ([#695](https://github.com/CarlsbergGBS/cx-component-library/issues/695)) ([45ddade](https://github.com/CarlsbergGBS/cx-component-library/commit/45ddadea01a0c56597032694df969fc5f2058559))
- **DSM-198:** Chips component ([#229](https://github.com/CarlsbergGBS/cx-component-library/issues/229)) ([c31055c](https://github.com/CarlsbergGBS/cx-component-library/commit/c31055c9adeab246b63af5af627d83bf04cbd60a))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-256:** link ([#279](https://github.com/CarlsbergGBS/cx-component-library/issues/279)) ([7185428](https://github.com/CarlsbergGBS/cx-component-library/commit/718542872ecffe4a580bfeb8bf71712317f8bad4))
- **DSM-288:** variants to input & chip ([#274](https://github.com/CarlsbergGBS/cx-component-library/issues/274)) ([5430764](https://github.com/CarlsbergGBS/cx-component-library/commit/54307643c5f171131601844690e7dd149de23086))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- New Chip component and minor changes ([#239](https://github.com/CarlsbergGBS/cx-component-library/issues/239)) ([ce92ab0](https://github.com/CarlsbergGBS/cx-component-library/commit/ce92ab0543b96296952a29e2d43d0df29d4e188f))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- **segmented-control:** [DSM-800] add tests ([#792](https://github.com/CarlsbergGBS/cx-component-library/issues/792)) ([09ee6fc](https://github.com/CarlsbergGBS/cx-component-library/commit/09ee6fc6a05da0bc1428306693bd763e3325680c))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
