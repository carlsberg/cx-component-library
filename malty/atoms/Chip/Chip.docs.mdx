---
labels: ['react', 'typescript', 'chip']
description: 'A Chip component.'
---
