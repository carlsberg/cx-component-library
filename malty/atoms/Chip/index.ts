export { Chip } from './Chip';
export { ChipSize } from './Chip.types';
export type { ChipProps } from './Chip.types';
