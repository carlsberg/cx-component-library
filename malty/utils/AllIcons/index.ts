export { allIcons, allIconsStoryOptions } from './AllIconsMap';
