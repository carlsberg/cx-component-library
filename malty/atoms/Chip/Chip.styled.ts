import styled, { css } from 'styled-components';
import { ChipSize } from './Chip.types';

export const StyledChip = styled.div<{
  height: string;
  hasButton?: boolean;
  selected: boolean;
  disabled?: boolean;
  readOnly?: boolean;
  size?: ChipSize;
}>`
  cursor: pointer;
  display: flex;
  align-items: center;
  height: ${({ height }) => height};
  width: fit-content;
  background-color: ${({ theme, selected }) =>
    selected ? theme.colors.colours.support[100].value : theme.colors.colours.support[20].value};
  padding: 0
    ${({ size, theme }) => {
      if (size === ChipSize.XSmall) {
        return theme.sizes['2xs'].value;
      }
      if (size === ChipSize.Small) {
        return theme.sizes.xs.value;
      }
      return theme.sizes.s.value;
    }};
  ${({ disabled, selected }) =>
    disabled &&
    css`
      pointer-events: none;
      cursor: default;
      background-color: ${({ theme }) => theme.colors.colours.system['disable-light-theme'].value};
      opacity: ${selected ? 1 : 0.75};
    `};
  ${({ readOnly, selected }) =>
    readOnly &&
    !selected &&
    css`
      pointer-events: none;
      cursor: default;
      background-color: ${({ theme }) => theme.colors.colours.support[20].value};
    `};
  ${({ readOnly, selected }) =>
    readOnly &&
    selected &&
    css`
      pointer-events: none;
      cursor: default;
      background-color: ${({ theme }) => theme.colors.colours.support[80].value};
    `};
  ${({ hasButton }) =>
    hasButton &&
    css`
      padding-right: 0;
    `};
  button {
    svg {
      transition: all 0.3s ease-in-out !important;
      ${({ selected }) =>
        selected &&
        css`
          transition: all 0.3s ease-in-out !important;
          transform: rotate(-45deg);
        `};
    }
  }
`;
export const StyledTextContainer = styled.div<{
  size?: ChipSize;
  hasIcon?: boolean;
  hasButton?: boolean;
}>`
  padding-left: ${({ size, theme, hasIcon }) => {
    if (hasIcon && size === ChipSize.XSmall) {
      return theme.sizes['4xs'].value;
    }
    return hasIcon && theme.sizes['2xs'].value;
  }};
  padding-right: ${({ hasButton, theme, size }) => {
    if (hasButton && size === ChipSize.XSmall) {
      return theme.sizes['2xs'].value;
    }
    if (hasButton && size === ChipSize.Small) {
      return theme.sizes.xs.value;
    }
    return hasButton && theme.sizes.s.value;
  }};
`;
export const StyledIcon = styled.div``;
