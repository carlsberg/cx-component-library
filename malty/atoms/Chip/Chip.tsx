/* eslint-disable no-nested-ternary */
import { Button, ButtonSize, ButtonStyle } from '@carlsberggroup/malty.atoms.button';
import { Icon, IconColor, IconName, IconSize } from '@carlsberggroup/malty.atoms.icon';
import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useEffect, useState } from 'react';
import { StyledChip, StyledIcon, StyledTextContainer } from './Chip.styled';
import { ChipProps, ChipSize } from './Chip.types';

export function Chip({
  label,
  size = ChipSize.Medium,
  selected = false,
  onChange = () => null,
  icon,
  showAction = false,
  dataTestId,
  disabled = false,
  readOnly = false,
}: ChipProps) {
  const theme = defaultTheme;
  const [chipSize, setChipSize] = useState(theme.sizes.l.value);
  const [fontSize, setFontSize] = useState(TextStyle.MediumSmallBold);
  const [buttonSize, setButtonSize] = useState(ButtonSize.Medium);

  const handleClick = () => {
    onChange(!selected);
  };

  useEffect(() => {
    switch (size) {
      case ChipSize.XSmall: {
        setChipSize(theme.sizes.m.value);
        setFontSize(TextStyle.SmallBold);
        setButtonSize(ButtonSize.XSmall);
        break;
      }
      case ChipSize.Small: {
        setChipSize(theme.sizes.l.value);
        setFontSize(TextStyle.MediumSmallBold);
        setButtonSize(ButtonSize.Small);
        break;
      }

      default: {
        setChipSize(theme.sizes.xl.value);
        setFontSize(TextStyle.MediumSmallBold);
        setButtonSize(ButtonSize.Medium);
        break;
      }
    }
  }, [size, theme]);
  return (
    <StyledChip
      hasButton={showAction}
      disabled={disabled}
      readOnly={readOnly}
      data-testid={dataTestId}
      onClick={handleClick}
      selected={selected}
      height={chipSize}
      size={size}
      theme={theme}
    >
      {!showAction && icon && (
        <StyledIcon data-testid={`${dataTestId}-icon`} theme={theme}>
          <Icon
            name={icon}
            color={
              readOnly && selected
                ? IconColor.White
                : readOnly && !selected
                ? IconColor.Support80
                : disabled
                ? IconColor.DisableLight
                : selected
                ? IconColor.White
                : IconColor.DigitalBlack
            }
            size={size === ChipSize.XSmall ? IconSize.Small : IconSize.MediumSmall}
          />
        </StyledIcon>
      )}

      <StyledTextContainer
        hasButton={showAction}
        hasIcon={!!icon}
        size={size}
        data-testid={`${dataTestId}-label`}
        theme={theme}
      >
        <Text
          textStyle={fontSize}
          color={
            (readOnly && selected
              ? TextColor.White
              : readOnly && !selected
              ? TextColor.Support80
              : disabled
              ? IconColor.DisableLight
              : selected
              ? IconColor.White
              : IconColor.DigitalBlack) as TextColor
          }
        >
          {label}
        </Text>
      </StyledTextContainer>
      {showAction && (
        <Button
          disabled={disabled || readOnly}
          negative={selected && !disabled}
          dataTestId={`${dataTestId}-button`}
          size={buttonSize}
          style={ButtonStyle.Transparent}
          icon={IconName.Plus}
        />
      )}
    </StyledChip>
  );
}
