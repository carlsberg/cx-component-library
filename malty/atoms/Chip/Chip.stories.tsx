import { Alert } from '@carlsberggroup/malty.icons.alert';
import { allIconsStoryOptions } from '@carlsberggroup/malty.utils.all-icons';
import { generateStorybookSpacing } from '@carlsberggroup/malty.utils.space';
import { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';
import { Chip } from './Chip';
import { ChipProps, ChipSize } from './Chip.types';

const ControlledChip = (props: ChipProps) => {
  const [stateChecked, setStateChecked] = useState(false);

  return <Chip {...props} selected={stateChecked} onChange={() => setStateChecked(!stateChecked)} />;
};

const meta: Meta<ChipProps> = {
  title: 'Forms/Chip',
  component: Chip,
  parameters: {
    importObject: 'Chip',
    importPath: '@carlsberggroup/malty.atoms.chip'
  },
  render: (args) => <ControlledChip {...args} />,
  argTypes: {
    label: {
      control: 'text',
      description: 'Label for the chip'
    },
    size: {
      description: 'Chip size. Options are',
      options: Object.values(ChipSize),
      table: {
        defaultValue: {
          summary: ChipSize.Medium
        },
        category: 'Styling'
      },
      control: 'select'
    },
    selected: {
      control: 'none',
      description: 'State of the component, selected or not selected',
      table: {
        category: 'State'
      }
    },
    showAction: {
      control: 'boolean',
      description: 'Displays add button if true'
    },
    onChange: {
      description: 'Function to be executed when Chip state changes',
      table: {
        category: 'Events'
      }
    },
    icon: {
      description: 'The icon component to be displayed',
      options: allIconsStoryOptions,
      control: 'select'
    },
    disabled: {
      control: 'boolean',
      description: 'Disable chip',
      table: {
        category: 'State'
      }
    },
    readOnly: {
      control: 'boolean',
      description: 'ReadOnly chip',
      table: {
        category: 'State'
      }
    },
    dataTestId: {
      control: 'text',
      description: 'Chip data-testid'
    },
    ...generateStorybookSpacing()
  }
};

type Story = StoryObj<ChipProps>;

export const Base: Story = {
  args: {
    label: 'Label',
    selected: false,
    showAction: false
  }
};

export const Button: Story = {
  args: {
    ...Base.args,
    showAction: true
  }
};

export const Icon: Story = {
  args: {
    ...Base.args,
    icon: <Alert />
  }
};

export const Selected: Story = {
  args: {
    ...Base.args,
    selected: true
  }
};

export const Disabled: Story = {
  args: {
    ...Base.args,
    selected: true,
    disabled: true
  }
};

export const ReadOnly: Story = {
  args: {
    ...Base.args,
    readOnly: true
  }
};

export default meta;
