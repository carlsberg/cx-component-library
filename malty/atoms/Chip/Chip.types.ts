import { IconName } from '@carlsberggroup/malty.atoms.icon';
import { SpaceProps } from '@carlsberggroup/malty.utils.space';

export interface ChipProps extends SpaceProps {
  selected: boolean;
  showAction?: boolean;
  label: string;
  size?: ChipSize;
  onChange?: (selected: boolean) => void;
  icon?: IconName;
  dataTestId?: string;
  disabled?: boolean;
  readOnly?: boolean;
}

export enum ChipSize {
  XSmall = 'XSmall',
  Small = 'Small',
  Medium = 'Medium'
}

export type StyledChipProps = {
  height: string;
  hasButton?: boolean;
  selected: boolean;
  disabled?: boolean;
  readOnly?: boolean;
  size?: ChipSize;
} & SpaceProps;
