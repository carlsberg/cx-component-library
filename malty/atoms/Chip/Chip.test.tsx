import { render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Chip } from './Chip';

const defaultLabel = 'label';

describe('chip', () => {
  it('renders with correct text', () => {
    render(<Chip selected={false} label={defaultLabel} />);
    expect(screen.getByText(defaultLabel)).toBeInTheDocument();
  });

  it('calls function onChange', () => {
    const onChange = jest.fn();
    render(<Chip selected={false} label={defaultLabel} onChange={onChange} />);
    fireEvent.click(screen.getByText(defaultLabel));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
