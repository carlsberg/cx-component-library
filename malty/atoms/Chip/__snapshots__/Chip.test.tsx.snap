// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chip matches snapshot 1`] = `
<span
  className="sc-bczRLJ hROJNc"
>
  <div
    className="sc-kDDrLX gyqSED"
    disabled={false}
    height="32px"
    onClick={[Function]}
    readOnly={false}
    selected={false}
    size="Medium"
  >
    <div
      className="sc-iqcoie fgUhqy"
      data-testid="undefined-label"
      size="Medium"
    >
      <span
        className="sc-bczRLJ hROJNc"
      >
        <p
          className="sc-dkzDqf cchzTL"
          color="digital-black"
        >
          label
        </p>
      </span>
    </div>
  </div>
</span>
`;
